File: java-code/src/Solns/binary_tree_zigzag_level_order_traversal_103.java
package Solns;

import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;



public class binary_tree_zigzag_level_order_traversal_103 {
	public static List<List<Integer>> zigzagLevelOrder(TreeNode root) {
		Queue<TreeNode> q = new LinkedList<>();

		List<List<Integer>> return_list = new LinkedList<List<Integer>>();

		if (root == null) {
			return return_list;
		}

		q.add(root);
		
		boolean odd = true;
		
		while(!q.isEmpty()) {
			int size = q.size();
			List<Integer> temp_list = new LinkedList<>();
			
			for(int i = 0; i < size; i++) {
				TreeNode currentNode = q.poll();
				
				if(odd)
					temp_list.add(currentNode.val);
				else
					temp_list.add(0, currentNode.val);
				
				if (currentNode.left != null)
					q.add(currentNode.left);
				if (currentNode.right != null)
					q.add(currentNode.right);
	
			}
		
			return_list.add(temp_list);
			if(odd)
				odd = false;
			else
				odd =true;
			
		}
		return return_list;
	}

	public static void main(String[] args) {
		// TODO Auto-generated method stub
		TreeNode root = new TreeNode(12);
		root.left = new TreeNode(7);
		root.right = new TreeNode(1);
		root.left.left = new TreeNode(9);
		root.right.left = new TreeNode(10);
		root.right.right = new TreeNode(5);

		System.out.println(zigzagLevelOrder(root));
	}
}

File: python-code/binary_tree_zigzag_level_order_traversal_103.py
class TreeNode:
  def __init__(self, val):
    self.val = val
    self.left, self.right = None, None


def zigzagLevelOrder(root):
    list = []

    if not root:
        return list

    queue = [root]

    even = False

    while queue:
        even = not even
        temp = []
        size = len(queue)

        for i in range(size):
            curr = queue.pop(0)

            if even:
                temp.append(curr.val)
            else:
                temp.insert(0, curr.val)

            if curr.left:

                queue.append(curr.left)
            if curr.right:
                queue.append(curr.right)

        list.append(temp)

    return list

def main():
    root = TreeNode(12)
    root.left = TreeNode(7)
    root.right = TreeNode(1)
    root.left.left = TreeNode(9)
    root.right.left = TreeNode(10)
    root.right.right = TreeNode(5)
    result = zigzagLevelOrder(root)

    print(result)

main()
